Basic PCAP parser to load into datastore

mnuzen opened the pull request:
# Description

- Allows for loading of the following PCAP attributes into datastore through parser with [pkts](https://github.com/aboutsip/pkts) library:
    - Source IP address
    - Destination IP address
    - Protocol (from list of protocols supported with pkts library) Type:
        - ICMP, IGMP, TCP, UDP, SCTP, SIP, SDP, ETHERNET_II, SLL, IPv4, RTP
    - Payload Size in bytes
    - Frequency 
        - Defined by number of packets exchanged between two IP addresses in PCAP file
- Reads PCAP files already loaded into repo with filename
- Includes unit testing for parsing of PCAP attributes to make sure correct elements are extracted

# Checklist:

- [x] Code follows the style guidelines
- [x] Code commented in relevant places
- [x] Changes to necessary documentation made
- [x] New and existing unit tests pass locally with changes

# Tests Performed:

- [x] Upload data to Datastore
- [x] Retrieve data from Datastore
- [x] Map Visualization 
- [x] Network Map Visualization 
- [x] List other manual test cases performed

mnuzen committed:
Refactoring packet parser

mnuzen committed:
Refactoring code

mnuzen committed:
Adding changes from network JS

mnuzen committed:
Removing vendor file

mnuzen committed:
Connecting to datastore

mnuzen committed:
Saving changes

mnuzen committed:
Connecting parser to datastore

mnuzen committed:
Working on grouping nodes

mnuzen committed:
Adding grouping in network graph

mnuzen committed:
Modifying physics

mnuzen committed:
Changing frequencies

mnuzen committed:
Refactoring code

mnuzen committed:
Resolving merge issues

mnuzen committed:
Re-adding parser

mnuzen committed:
Removing printing packets functionality

mnuzen committed:
Modularizing packet parser

mnuzen committed:
Adding more PCAP files

mnuzen committed:
Added frequency functionality for parser

mnuzen committed:
Resolving sr/dst IP's

mnuzen committed:
Adding protocol detection

mnuzen committed:
Adding arrows

mnuzen committed:
Modifying to incorporate new datastore changes

mnuzen committed:
Commenting parser

mnuzen committed:
Merge remote-tracking branch 'origin/Erik_Testing' into melba_parser

mnuzen committed:
Incorporating changes

arunk054 commented on src/main/java/com/google/sps/datastore/GenericPCAPDao.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface GenericPCAPDao {
+
+    public ArrayList<PCAPdata> getPCAPObjects(String searchEntity);
```
I thought you already had this thing in another Pull request from Erik.

arunk054 commented on src/main/java/com/google/sps/servlets/PacketParserServlet.java:
```diff
@@ -0,0 +1,167 @@
+  //static final String FILENAME = "WEB-INF/files/chargen-tcp.pcap";
+  static final String FILENAME = "WEB-INF/files/smallFlows.pcap";
+
+  static final String MYIP = "192.168.3.131"; // for smallFlows.pcap 
```
This is complicating things a bit. The role of the parser should be to only "parse" the file and avoid as much processing as possible. Since you are early in your development if you can break it down into two phases that would be nice. I understand the fundamental challenge in doing so is to update the 'frequency' in the entity after it has been parsed. For now, you can leave that as it is, i.e. you will still be updating the Entity with pcap data and frequency but you can have two functions. The first one just parses the data into a datastructure. The second will compute the frequencies and a datastructure  linking each row  to frequency. You will then use both these to write to the datastore.

arunk054 commented on src/main/java/com/google/sps/servlets/PacketParserServlet.java:
```diff
@@ -0,0 +1,167 @@
+        String srcip = ip.getSourceIP();
+        String dstip = ip.getDestinationIP();
+
+        if (srcip == MYIP) {
```
you are combining the processing in the parsing. I know it is efficient to do that while you are parsing but that is a trade-off worth since parsing is several orders of magnitude more time than looping the array again. So try to see if you can decouple this.

arunk054 commented on src/main/java/com/google/sps/servlets/PacketParserServlet.java:
```diff
@@ -0,0 +1,167 @@
+  /* Determine the protocol of IPv4 packet; if unknown, then returns "IPv4" as protocol.
+    * Checks for the following protocols — ICMP, IGMP, TCP, UDP, SCTP, SIP, SDP, ETHERNET_II, SLL, IPv4, RTP —
+    * which are the only protocols supported by pkts library. */
+  public String getProtocol(IPPacket packet) throws IOException {
```
This should be a private method.

arunk054 commented on src/main/java/com/google/sps/servlets/PacketParserServlet.java:
```diff
@@ -0,0 +1,167 @@
+  static final String FILENAME = "WEB-INF/files/smallFlows.pcap";
+
+  static final String MYIP = "192.168.3.131"; // for smallFlows.pcap 
+  static final String ENTITY_TAG = "file_1";
```
Start out with passing the file and my_IP as parameters to your servlet, that extra work will make things easier for you.

mnuzen replied:
Resolved in latest commits.

arunk054 commented on src/main/java/com/google/sps/servlets/PacketParserServlet.java:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  /* Reads PCAP file (from file name) as a stream and puts unique connections into allPCAP HashMap. */
+  public void parseRaw(String file) throws IOException {
```
Avoid having these functions in your servlet itself. Create separate class called PCAPParser that takes in the file-name in constructor. That will then have a function parseRaw that returns a list of PCAP objs or creates it. That class can also  have the write PCAP to datastore function.

mnuzen replied:
Resolved in latest commits.

mnuzen committed:
Merge remote-tracking branch 'origin/map_viz' into melba_parser

mnuzen committed:
Resolving merge conflict

mnuzen committed:
Adding in PCAP file selection functionality

mnuzen committed:
Adding in parser interface

mnuzen commented:
- Added separate PCAP Parser class (so parser servlet calls those class functions)
- Changed loading functionality so that file name and local IP address are passed into the parser
- Added input options on `/packet.html` file for user input of file name and local IP address

mnuzen committed:
Merge remote-tracking branch 'origin/master' into melba_parser

mnuzen committed:
Obfuscating files and decoupling parsing from processing

mnuzen committed:
Changing frequency incrementer

mnuzen committed:
Refactoring code

mnuzen committed:
Merge remote-tracking branch 'origin/ci-testing' into melba_parser

mnuzen committed:
Instructions for sample file

mnuzen committed:
Adding parser tests

mnuzen committed:
Adding parser tests

mnuzen committed:
Merge remote-tracking branch 'origin/ci-testing' into melba_parser

mnuzen committed:
Incorporating new CI changes

mnuzen committed:
Changing yaml

mnuzen committed:
yaml

mnuzen committed:
yaml

mnuzen committed:
test

mnuzen committed:
test

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java:
```diff
@@ -0,0 +1,166 @@
+  private ArrayList<PCAPdata> allPCAP = new ArrayList<PCAPdata>(); 
+  private HashMap<String, PCAPdata> finalPCAP = new HashMap<String, PCAPdata>();
+
+  private String FILENAME;
```
only constants are ALL CAPS. like public static final String FILENAME

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java:
```diff
@@ -0,0 +1,166 @@
+        String protocol = getProtocol(ip); 
+
+        if (first) {
+            MYIP = srcip;
```
can you comment on what is the MYIP for.

mnuzen replied:
An explanation's above the class but I can add it again here

arunk054 replied:
Ah thanks, I missed it. yes typically the comments are made right above the variable declaration or where you use it.

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java:
```diff
@@ -0,0 +1,166 @@
+      String srcip = packet.source;
+      String dstip = packet.destination;
+
+      if (srcip == MYIP) {
```
This if condition will always be true right since you assign the MYIP = srcip above?  It is not very clear what are we trying to do here with the MYIP?

mnuzen replied:
Since MYIP is the local address and OUTIP is the "outside" machine, in the following packets the order could be flipped where (MYIP, OUTIP) are outgoing packets and (OUTIP, MYIP) are incoming packets; we're using OUTIP as the key in the hashmap since that's always unique.

arunk054 replied:
Thanks for the clarification

mnuzen committed:
testing stream

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java:
```diff
@@ -0,0 +1,166 @@
+        currPCAP.incrementFrequency();
+      }
+      else {
+        PCAPdata tempPCAP = new PCAPdata(MYIP, OUTIP, "", "", packet.protocol, packet.size, packet.flagged, packet.frequency); 
```
if this temp is the same as `packet` may be  just add it to `finalPCAP.put(OUTIP, packet)`

mnuzen committed:
testing yaml

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java:
```diff
@@ -0,0 +1,166 @@
+      if (finalPCAP.containsKey(OUTIP)){
+        // retrieve current value with OUTIP and increments frequency
+        PCAPdata currPCAP = finalPCAP.get(OUTIP);
+        currPCAP.incrementFrequency();
```
This looks like you are counting frequencies as directional?

mnuzen replied:
I think the frequency is incremented regardless of the direction since the packets in the hashmap correspond to the "outside" address always

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java:
```diff
@@ -0,0 +1,166 @@
+    return protocol;
+  }
+ 
+  /* Removes duplicated data and increments corresponding frequencies */
```
In the process of removing duplicates are you not accidentally removing other information? You are assuming that between a pair of SRC and DEST the protocol, size etc will also be same, which may not be true.

arunk054 replied:
The fundamental problem this is getting complicated is because you are tying frequency to each PCAP data row, whereas if you had just the PCAP data as it is, then you can store the frequencies separately in a entity something like: `srcip, destip, frequency`. The ip-pair could be be directional and if you want to aggregate them into undirectional that should be fairly easy if you loop through each unique `srcip` in the list and check if it is in a `destip`, then increment the map where that row's srcip is in current rows destip.

arunk054 commented on src/main/tests/dao/PCAPParserDaoImplTest.java:
```diff
@@ -0,0 +1,109 @@
+/* Compares file2.pcap's parsed results to hard-coded results. IP1 and IP2 addresses are located in hidden text file and retrieved as a stream. */
+public class PCAPParserDaoImplTest {
+  public PCAPParserDao parser;
+  public String PCAPNAME = "files/file2.pcap";
```
Are these files generated or public ones.

mnuzen replied:
These are public ones, but the actual files are hidden (and accessed through class loader).

arunk054 commented on src/main/tests/dao/PCAPParserDaoImplTest.java:
```diff
@@ -0,0 +1,109 @@
+  @Test
+  public void testFrequency() {
+    HashMap<String, PCAPdata> finalPCAP = parser.getFinalPCAP();
+    int frequency = finalPCAP.get(IP2).getFrequency();
```
also add asserts for the number of rows here after removing duplicates and above with duplicates.

arunk054 reviewed (approved):
Looks good. But we might need one more iteration at decoupling the frequency out of the PCAPdata

mnuzen committed:
Merge remote-tracking branch 'origin/master' into melba_parser

mnuzen committed:
testing

mnuzen committed:
Updating commons version

mnuzen committed:
Modifying file paths

mnuzen committed:
Resolving comments

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java:
```diff
@@ -0,0 +1,173 @@
+   * allPCAP HashMap indexes all unique outips in a PCAP file to a PCAPdata object that stores needed attributes, 
+     including the number of connections between myip and outip 
+   * We load datastore with all of the unique PCAPdata objects from our HashMap with a given entity tag.
+   * Hard-coded the source IP address (myip) -- need to find a way to retrieve myip (look at first packet, ask user, use Whatsmyip?)
```
This is something we should ask Philip on the chat room. Generally trying to extract user's IP from the PCAP file sounds intrusive or even asking them.
What about the public pcap files. Is it possible to use a simple heuristic like check the first few rows and determine which is the common ip and set that as the myip. Or you can do something fancy like learning on the fly, i.e. as you read more rows you can correct the assumption.
Also if you only have 2 IPs in your capture it will be impossible to tell i guess. So you could just store them as directional being agnostic of MYIP. Then yo can always compute the undirectional frequency from directional data.

arunk054 reviewed (approved):
This is ok for MVP but I think we should take another look at the way we parse IPs and store frequencies.

mnuzen committed:
test

mnuzen commented:
> This is ok for MVP but I think we should take another look at the way we parse IPs and store frequencies.

Resolved in #49
